Branch-SDK/src/io/branch/referral/Branch.java

```diff
@@ -967,27 +1015,57 @@ private void generateShortLink(final String alias, final int type, final Collect
 			}
 			
 			if (linkCache_.containsKey(linkPost)) {
+				String url = linkCache_.get(linkPost);
 				if (linkCreateCallback_ != null) {
-					linkCreateCallback_.onLinkCreate(linkCache_.get(linkPost), null);
+					linkCreateCallback_.onLinkCreate(url, null);
 				}
+				return url;
 			} else {
-				new Thread(new Runnable() {
-					@Override
-					public void run() {
-						ServerRequest req = new ServerRequest(BranchRemoteInterface.REQ_TAG_GET_CUSTOM_URL, linkPost);
-						if (!initFailed_) {
-							requestQueue_.enqueue(req);
-						}
-						if (initFinished_ || !hasNetwork_) {
-							lastRequestWasInit_ = false;
-							processNextQueueItem();
-						} else if (initFailed_ || initNotStarted_) {
-							handleFailure(req);
-						}
-					}
-				}).start();
+				ServerRequest req = new ServerRequest(BranchRemoteInterface.REQ_TAG_GET_CUSTOM_URL, linkPost);
+				if (async) {
+					generateShortLinkAsync(req);
+				} else {
+					return generateShortLinkSync(req);
+				}
 			}
 		}
+		return null;
+	}
+	
+	private String generateShortLinkSync(ServerRequest req) {
+		if (hasUser()) {
+			ServerResponse response = kRemoteInterface_.createCustomUrlSync(req.getPost());
+			String url = prefHelper_.getUserURL();
+			if (response.getStatusCode() == 200) {
+				try {
+					url = response.getObject().getString("url");
+					linkCache_.put(response.getLinkData(), url);
```

Comment:
For safety sake, I was checking if response.getLinkData() is null first before cacheing the link.